core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1552,4 +1582,65 @@ private[spark] object BlockManager {
     override val metricRegistry = new MetricRegistry
     metricRegistry.registerAll(metricSet)
   }
+
+  class RemoteBlockTempFileManager(blockManager: BlockManager)
+      extends TempFileManager with Logging {
+
+    private class ReferenceWithCleanup(file: File, referenceQueue: JReferenceQueue[File])
+        extends WeakReference[File](file, referenceQueue) {
+      private val filePath = file.getAbsolutePath
+
+      def cleanUp(): Unit = {
+        logDebug(s"Clean up file $filePath")
+
+        if (!new File(filePath).delete()) {
+          logDebug(s"Fail to delete file $filePath")
+        }
+      }
+    }
```

Comment:
Yes, it's good idea to delete the tmp files by `WeakReference`. Is it possible to make this a default behavior of `TempFileManager`, and make it to be `override def registerTempFileToClean(file: File): Unit`.
In https://github.com/apache/spark/pull/18565, I was struggling for file leak, because `TempFileManager` are not responsible for deleting tmp files.
Additionally, files are deleted by the `cleaningThread`, hope the cost can be neglected.

Reply:
I think the overhead is not big, but I'm not sure why there's a file leak issue here in `ShuffleBlockFetcherIterator` (the implementation of `TempFileManager`). From the code, all the temp files are tracked in `shuffleFilesSet`, and will be deleted during `cleanup`, can you please elaborate more?

Reply:
Yeah, the thing is if we `registerTempFileToClean` after `cleanup`, the tmp file will never be cleaned. In #18565, if `registerTempFileToClean` return false, it means the `TempFileManager` has already stopped, i.e. `cleanup` has already been finished and will not be called again. We can not rely on `TempFileManager` to clean the tmp file.

Reply:
But here in `ShuffleBlockFetcherIterator#registerTempFileToClean`, the caller will delete the file if it returns false, does it still has file leak problem?

Reply:
I don't thinks so:
if `false`, `ShuffleBlockFetcherIterator` is stopped and caller remove file;
if `true`, `ShuffleBlockFetcherIterator` remove file.

Reply:
Yes, that's what I mean. No matter false (the caller) or true (`ShuffleBlockFetcherIterator`), the file will be deleted, that's my question why there still has file leak issue?

Reply:
Ah, I just think caller should be very careful for file leak. Maybe we can delete by weak reference. So caller just `registerTempFileToClean` and don't have to cleanup file itself and no worry about the file leak.
I just think the implementation in this patch is good to void potential file leak.

Reply:
My concern is that: for shuffle part, since there's a explicit API to `cleanup` temp files, so it's not so necessary to track again with weak reference. Also weak reference is triggered with GC, and shuffle operations are usually much more frequent and heavier, using weak reference to track temp shuffle files may increase the overhead of GC probably. Whereas, compared to shuffle, fetching remote blocks are happened occasionally when block is not cached in local, so using weak reference may not increase the overhead a lot.

Reply:
I get it, thanks a lot for explanation :)